src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +280,144 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
+            String fieldType = getFieldTypeFromMapping(index, fieldName, allIndicesMap.get(index.getName()));
+            if (fieldType != null) {
+                cacheFieldType(index, fieldName, fieldType);
+                return fieldType;
+            }
+        }
+        return null;
+    }
+
+    String getFieldTypeFromMapping(Index index, String fieldName, MappingMetadata mappingMetadata) {
+        // Get properties from cache or mapping metadata
+        Map<String, ?> propertiesAsMap = propertiesCache.computeIfAbsent(index, k -> {
+            if (mappingMetadata != null) {
+                return (Map<String, ?>) mappingMetadata.getSourceAsMap().get("properties");
+            }
+            return null;
+        });
+
+        // Recursively find the field type from properties
+        if (propertiesAsMap != null) {
+            String fieldType = findFieldTypeInProperties(propertiesAsMap, fieldName.split("\\."), 0);
+
+            // Cache the NO_FIELD_TYPE_VALUE result if no field type is found in this index
+            if (fieldType == null) {
+                cacheFieldType(index, fieldName, NO_FIELD_TYPE_VALUE);
+            }
+            return fieldType;
+        }
+        return null;
+    }
+
+    private void cacheFieldType(Index index, String fieldName, String fieldType) {
+        fieldTypeMap.computeIfAbsent(index, k -> new ConcurrentHashMap<>()).put(fieldName, fieldType);
+    }
+
+    private String findFieldTypeInProperties(Map<String, ?> properties, String[] fieldParts, int index) {
+        if (index >= fieldParts.length) {
+            return null;
+        }
+
+        String currentField = fieldParts[index];
+        Object currentMapping = properties.get(currentField);
+
+        // Check if current mapping is a map and contains further nested properties or a type
+        if (currentMapping instanceof Map) {
+            Map<String, ?> currentMappingMap = (Map<String, ?>) currentMapping;
+
+            // If it has a 'properties' key, go into the nested object
+            if (currentMappingMap.containsKey("properties")) {
+                Map<String, ?> nestedProperties = (Map<String, ?>) currentMappingMap.get("properties");
+                return findFieldTypeInProperties(nestedProperties, fieldParts, index + 1);
+            }
+
+            // If it has a 'type' key, return the type
+            if (currentMappingMap.containsKey("type")) {
+                return currentMappingMap.get("type").toString();
+            }
+        }
+
+        return null;
+    }
+
+    private Map<String, MappingMetadata> getStringMappingMetadataMap(Set<Index> successfulSearchShardIndices) {
+        try {
+            return clusterService.state()
+                .metadata()
+                .findMappings(successfulSearchShardIndices.stream().map(Index::getName).toArray(String[]::new), input -> (str -> true));
+        } catch (IOException e) {
+            return null;
```

Comment:
We should instead use `clusterService.state().metadata().index(index)` once we start doing iterating over indices from the top method